source-firestore/main.go

```diff
@@ -0,0 +1,411 @@
+		result.Message = err.Error()
+	}
+	defer client.Close()
+
+	return airbyte.NewStdoutEncoder().Encode(airbyte.Message{
+		Type:             airbyte.MessageTypeConnectionStatus,
+		ConnectionStatus: result,
+	})
+}
+
+const DISCOVER_DOC_LIMIT = 40000
+
+func doDiscover(args airbyte.DiscoverCmd) error {
+	var cfg config
+	if err := args.ConfigFile.Parse(&cfg); err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	sa := option.WithCredentialsJSON([]byte(cfg.CredentialsJSON))
+	app, err := firebase.NewApp(ctx, nil, sa)
+	if err != nil {
+		return err
+	}
+
+	client, err := app.Firestore(ctx)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	var catalog = new(airbyte.Catalog)
+	var collections = client.Collections(ctx)
+	for {
+		var collection, err = collections.Next()
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		var query = collection.Query.Limit(DISCOVER_DOC_LIMIT)
+		var docs = query.Documents(ctx)
+
+		var docsCh = make(chan schema_inference.Document)
+		var logEntry = log.WithField("collection", collection.ID)
+
+		var eg = new(errgroup.Group)
+		var schema schema_inference.Schema
+
+		eg.Go(func() error {
+			schema, err = schema_inference.Run(ctx, logEntry, docsCh)
+			if err != nil {
+				return fmt.Errorf("schema inference: %w", err)
+			}
+			return nil
+		})
+
+		for {
+			var doc, err = docs.Next()
+			if err == iterator.Done {
+				break
+			} else if err != nil {
+				return err
+			}
+			var data = doc.Data()
+			data[firestore.DocumentID] = doc.Ref.ID
+			docJson, err := json.Marshal(data)
+			docsCh <- docJson
+		}
+		close(docsCh)
+
+		if err := eg.Wait(); err != nil {
+			return err
+		}
+
+		catalog.Streams = append(catalog.Streams, airbyte.Stream{
+			Name:                    collection.ID,
+			JSONSchema:              schema,
+			SupportedSyncModes:      []airbyte.SyncMode{airbyte.SyncModeIncremental},
+			SourceDefinedCursor:     true,
+			SourceDefinedPrimaryKey: [][]string{{firestore.DocumentID}},
+		})
+	}
+
+	var encoder = airbyte.NewStdoutEncoder()
+	return encoder.Encode(airbyte.Message{
+		Type:    airbyte.MessageTypeCatalog,
+		Catalog: catalog,
+	})
+}
+
+// struct sent over channel to encoder
+type encDocument struct {
+	doc        []byte
+	streamName string
+	state      []byte
+}
+
+func doRead(args airbyte.ReadCmd) error {
+	var cfg config
+	var state State
+	var catalog airbyte.ConfiguredCatalog
+
+	if err := args.ConfigFile.Parse(&cfg); err != nil {
+		return err
+	} else if err := args.CatalogFile.Parse(&catalog); err != nil {
+		return err
+	} else if args.StateFile != "" {
+		if err := args.StateFile.Parse(&state); err != nil {
+			return err
+		}
+	}
+
+	var lastScan = time.UnixMilli(0)
+	if state.LastScan != "" {
+		var err error
+		lastScan, err = time.Parse(time.RFC3339, state.LastScan)
+		if err != nil {
+			return fmt.Errorf("parsing last scan checkpoint failed: %w", err)
+		}
+	}
+
+	var scanInterval, _ = time.ParseDuration("12h")
```

Comment:
Nit: It would nice if the default `12h` value was declared as a constant, and maybe set in `(c *config) Validate()` if there isn't one supplied by the user. Took me a minute to figure out how the default was set.